src/remap/function/parse_grok.rs

```diff
@@ -0,0 +1,149 @@
+use remap::prelude::*;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseGrok;
+
+impl Function for ParseGrok {
+    fn identifier(&self) -> &'static str {
+        "parse_grok"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::String(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "pattern",
+                accepts: |v| matches!(v, Value::String(_)),
+                required: true,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required_expr("value")?;
+
+        let patternstr = match arguments
+            .required_expr("pattern")?
+            .literal()
+            .ok_or("grok pattern must be a literal string")?
+        {
+            Value::String(patternstr) => String::from_utf8_lossy(&patternstr).into_owned(),
+            _ => {
+                // The parameter type check should prevent us from reaching this point.
+                return Err(Error::Unknown);
+            }
+        };
+
+        let mut grok = grok::Grok::with_patterns();
+        let pattern = Arc::new(
+            grok.compile(&patternstr, true)
+                .map_err(|e| Error::from(e.to_string()))?,
+        );
+
+        Ok(Box::new(ParseGrokFn { value, pattern }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseGrokFn {
+    value: Box<dyn Expression>,
+    // Wrapping pattern in an Arc, as cloning the pattern could otherwise be expensive.
+    pattern: Arc<grok::Pattern>,
+}
+
+impl ParseGrokFn {
+    #[cfg(test)]
+    fn new(value: Box<dyn Expression>, pattern: String) -> Result<Self> {
+        let mut grok = grok::Grok::with_patterns();
+        let pattern = Arc::new(
+            grok.compile(&pattern, true)
+                .map_err(|e| Error::from(e.to_string()))?,
```

Comment:
Since `?` does `Into::into` behind the scenes, would this work as well?

```suggestion
            grok.compile(&pattern, true).map_err(ToString::to_string)?,
```

Reply:
Unfortunately, `to_string` wants the reference to `e`, rather than just `e`, so the best I can do seems to be: 

```
                .map_err(|e| e.to_string())?,
```